mountpoint-s3-client/src/imds_crt_client.rs

```diff
@@ -56,32 +50,30 @@ impl ImdsCrtClient {
     }
 
     /// Query the type of the EC2 instance this code is executed on.
-    pub fn make_instance_type_query(&self) -> Result<ImdsQueryRequest, ImdsQueryRequestError> {
+    pub async fn get_instance_type(&self) -> Result<String, ImdsQueryRequestError> {
         let (tx, rx) = oneshot::channel();
         self.imds_client.get_instance_type(move |result| {
-            let _ = tx.send(result.map_err(ImdsQueryRequestError::CrtError));
+            let _ = tx.send(result);
         })?;
 
-        Ok(ImdsQueryRequest { receiver: rx })
+        match rx.await {
+            Ok(result) => result.map_err(ImdsQueryRequestError::CrtError),
+            Err(err) => Err(ImdsQueryRequestError::InternalError(Box::new(err))),
+        }
     }
-}
 
-#[derive(Debug)]
-#[pin_project]
-/// ImdsQueryRequest is a data encapsulation wrapper for asnychronous instance metadata query.
-pub struct ImdsQueryRequest {
-    #[pin]
-    receiver: oneshot::Receiver<Result<String, ImdsQueryRequestError>>,
-}
-
-impl Future for ImdsQueryRequest {
-    type Output = Result<String, ImdsQueryRequestError>;
+    /// Query the region of the EC2 instance this code is executed on.
+    pub async fn get_region(&self) -> Result<String, ImdsQueryRequestError> {
```

Comment:
So the test failures you're getting are because the tests are hitting Azure's IMDS, which listens on the same IP and I guess is where GitHub Actions runs. That's an unusual thing but we probably shouldn't totally choke. Maybe some cheap regex check on the string, like `^[A-Za-z0-9-]+$`, is enough to get only valid-ish regions without overfitting.

Reply:
Yes, maybe we should do that. But I'm not sure I understand why that string is returned:
* locally I get an empty string, but the status code is 403
* github actions, I *guess* the status code is 404
It seems that the crt does not check the status code before calling our callback.